builder/dockerfile/internals.go

```diff
@@ -166,8 +166,12 @@ func (b *Builder) performCopy(state *dispatchState, inst copyInstruction) error
 	runConfigWithCommentCmd := copyRunConfig(
 		state.runConfig,
 		withCmdCommentString(commentStr, state.operatingSystem))
-	hit, err := b.probeCache(state, runConfigWithCommentCmd)
-	if err != nil || hit {
+
+	if hit, err := b.probeCache(state, runConfigWithCommentCmd); err != nil || hit {
+		return err
+	}
+
+	if _, err := b.create(runConfigWithCommentCmd); err != nil {
```

Comment:
So this is the actual fix, correct? There's also quite some refactoring in this PR (See my other comment), and I'd prefer the actual fix to be minimal (so that it's clear where the fix is).

Trying what the minimal change is,  I came down to:

```patch
diff --git a/builder/dockerfile/internals.go b/builder/dockerfile/internals.go
index 53748f0619..31822bcfa2 100644
--- a/builder/dockerfile/internals.go
+++ b/builder/dockerfile/internals.go
@@ -166,8 +166,7 @@ func (b *Builder) performCopy(state *dispatchState, inst copyInstruction) error
 	runConfigWithCommentCmd := copyRunConfig(
 		state.runConfig,
 		withCmdCommentString(commentStr, state.operatingSystem))
-	hit, err := b.probeCache(state, runConfigWithCommentCmd)
-	if err != nil || hit {
+	if id, err := b.probeAndCreate(state, runConfigWithCommentCmd); err != nil || id == "" {
 		return err
 	}
 
```

With that patch:

```
make BIND_DIR=. binary shell

make install
dockerd &> /var/log/docker.log &

mkdir /foo && cd /foo

cat <<EOF> Dockerfile
FROM alpine
COPY Dockerfile /
ADD Dockerfile.* /
RUN echo "hello world"
EOF

cp Dockerfile Dockerfile.bla

docker build --no-cache -t copy .
```

Which produces the output below (and is what I think is addressing the issue - please double check):

````
Sending build context to Docker daemon  3.072kB
Step 1/4 : FROM alpine
latest: Pulling from library/alpine
ff3a5c916c92: Pull complete 
Digest: sha256:e1871801d30885a610511c867de0d6baca7ed4e6a2573d506bbec7fd3b03873f
Status: Downloaded newer image for alpine:latest
 ---> 3fd9065eaf02
Step 2/4 : COPY Dockerfile /
Removing intermediate container ee6df06eaa3e
 ---> 2b40e2810b52
Step 3/4 : ADD Dockerfile.* /
Removing intermediate container bf108809cf1c
 ---> e0d9c73e400c
Step 4/4 : RUN echo "hello world"
 ---> Running in cffffaeab0ff
hello world
Removing intermediate container cffffaeab0ff
 ---> 90f6265e7b5c
Successfully built 90f6265e7b5c
Successfully tagged copy:latest
```

**However**, I'm wondering if the container create is only for "cosmetic" purposes (i.e., just to show "removing intermediate container");

- Is there an actual purpose for this container?
- If not; perhaps this was no longer done (as an optimisation step)

If it's not needed, I think we should not make this change, and accept that the output changed?